python-recipes/agents/03_memory_agent.ipynb

```diff
@@ -0,0 +1,1647 @@
+    "    while True:\n",
+    "        user_input = input(\"\\nYou (type 'quit' to quit): \")\n",
+    "\n",
+    "        if not user_input:\n",
+    "            continue\n",
+    "\n",
+    "        if user_input.lower() in [\"exit\", \"quit\"]:\n",
+    "            print(\"Thank you for using the Travel Assistant. Goodbye!\")\n",
+    "            break\n",
+    "\n",
+    "        state[\"messages\"].append(HumanMessage(content=user_input))\n",
+    "\n",
+    "        try:\n",
+    "            # Process user input through the graph\n",
+    "            for result in graph.stream(state, config=config, stream_mode=\"values\"):\n",
+    "                state = RuntimeState(**result)\n",
+    "\n",
+    "            logger.debug(f\"# of messages after run: {len(state['messages'])}\")\n",
+    "\n",
+    "            # Find the most recent AI message, so we can print the response\n",
+    "            ai_messages = [m for m in state[\"messages\"] if isinstance(m, AIMessage)]\n",
+    "            if ai_messages:\n",
+    "                message = ai_messages[-1].content\n",
+    "            else:\n",
+    "                logger.error(\"No AI messages after run\")\n",
+    "                message = \"I'm sorry, I couldn't process your request properly.\"\n",
+    "                # Add the error message to the state\n",
+    "                state[\"messages\"].append(AIMessage(content=message))\n",
+    "\n",
+    "            print(f\"\\nAssistant: {message}\")\n",
+    "\n",
+    "            # Add the current state to the memory processing queue\n",
+    "            if memory_strategy == MemoryStrategy.MANUAL:\n",
+    "                memory_queue.put((state.copy(), config))\n",
+    "\n",
+    "        except Exception as e:\n",
+    "            logger.exception(f\"Error processing request: {e}\")\n",
+    "            error_message = \"I'm sorry, I encountered an error processing your request.\"\n",
+    "            print(f\"\\nAssistant: {error_message}\")\n",
+    "            # Add the error message to the state\n",
+    "            state[\"messages\"].append(AIMessage(content=error_message))\n",
+    "\n",
+    "\n",
+    "try:\n",
+    "    user_id = input(\"Enter a user ID: \") or \"demo_user\"\n",
+    "    thread_id = input(\"Enter a thread ID: \") or \"demo_thread\"\n",
+    "except Exception:\n",
+    "    # If we're running in CI, we don't have a terminal to input from, so just exit\n",
+    "    exit()\n",
+    "else:\n",
+    "    main(thread_id, user_id)\n"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "env",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.11.11"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
```

Comment:
Awesome stuff here -- let's make sure we summarize all of the components in either a conclusion/summary and/or diagram with maybe a link to some CTA or next step?

Reply:
Good call, I'll link to LangGraph and our Redis checkpoint repo.